src/rules/trailingCommaRule.ts

```diff
@@ -62,205 +67,106 @@ export class Rule extends Lint.Rules.AbstractRule {
-        const children = node.getChildren();
-        for (let i = 0; i < children.length - 2; i++) {
-            if (children[i].kind === ts.SyntaxKind.OpenBraceToken &&
-                children[i + 1].kind === ts.SyntaxKind.SyntaxList &&
-                children[i + 2].kind === ts.SyntaxKind.CloseBraceToken) {
-                const grandChildren = children[i + 1].getChildren();
-                // the AST is different from the grammar spec. The semicolons are included as tokens as part of *Signature,
-                // as opposed to optionals alongside it. So instead of children[i + 1] having
-                // [ PropertySignature, Semicolon, PropertySignature, Semicolon ], the AST is
-                // [ PropertySignature, PropertySignature], where the Semicolons are under PropertySignature
-                const hasSemicolon = grandChildren.some((grandChild) =>
-                    Lint.childOfKind(grandChild, ts.SyntaxKind.SemicolonToken) !== undefined);
-
-                if (!hasSemicolon) {
-                    const endLineOfClosingElement = this.getSourceFile().getLineAndCharacterOfPosition(children[i + 2].getEnd()).line;
-                    this.lintChildNodeWithIndex(children[i + 1], grandChildren.length - 1, endLineOfClosingElement);
-                }
+        const sourceText = this.sourceFile.text;
+        for (const member of members) {
+            // PropertSignature in TypeLiteral can end with semicolon or comma. If one ends with a semicolon, don't check for trailing comma
+            if (sourceText[member.end - 1] === ";") {
+                return;
             }
         }
-        super.visitTypeLiteral(node);
-    }
-
-    public visitTypeReference(node: ts.TypeReferenceNode) {
-        this.lintNode(node);
-        super.visitTypeReference(node);
+        // The trailing comma is part of the last member and therefore not present as hasTrailingComma on the NodeArray
+        const hasTrailingComma = sourceText[members.end - 1] === ",";
+        return this.checkComma(hasTrailingComma, members, node.end);
     }
 
-    private lintNode(node: ts.Node, includeBraces?: boolean) {
-        const children = node.getChildren();
-        const syntaxListWrapperTokens = (includeBraces === true) ?
-            TrailingCommaWalker.SYNTAX_LIST_WRAPPER_TOKENS : TrailingCommaWalker.SYNTAX_LIST_WRAPPER_TOKENS.slice(1);
-
-        for (let i = 0; i < children.length - 2; i++) {
-            syntaxListWrapperTokens.forEach(([openToken, closeToken]) => {
-                if (children[i].kind === openToken &&
-                    children[i + 1].kind === ts.SyntaxKind.SyntaxList &&
-                    children[i + 2].kind === closeToken) {
-                    this.lintChildNodeWithIndex(node, i + 1);
-                }
-            });
+    private checkListWithEndToken(node: ts.Node, list: ts.NodeArray<ts.Node>, closeTokenKind: ts.SyntaxKind) {
```

Comment:
I'd prefer to leave it as is.
Maybe someday type parameters can have trailing commas. Then this method could also handle `GreaterThenToken`.